Added the ppolicy module and logic to add ldifs

SachinKSingh28 opened the pull request:
This PR is to add some password policies to adop ldap. we are using ppolicy module.

To review this PR -

- Add following environment variables in the compose section for ldap -
`
    SLAPD_ADDITIONAL_SCHEMAS: ppolicy
    SLAPD_ADDITIONAL_MODULES: ppolicy
`
- If the LDAP container is already running then add another parameter -
`
SLAPD_FORCE_RECONFIGURE: true
`

- Run compose.

SachinKSingh28 committed:
Added the ppolicy module and logic to add ldifs

RobertNorthard commented:
Thanks @SachinKSingh28 

Default password policy.

- [x] Password Attribute: userPassword
- [x] Password failure count interval: 30 **Can we increase the interval time? To slow down brute-forcing attacks even more e.g. 15 minutes**
- [x] Password Grace Auth N Limit: 5 **(This gives the user 5 opportunities once the password has expired to reset)**
- [x] Password in history: 5 **(store previous 'x' number of passwords)
- [x] Password Lockout: TRUE
- [x] Password Lockout duration: 1800 (30 minutes)
- [x] Password Max Age: 15552000 (180 days) - **I think 90 days is better**
- [x] Password Max Failures: 5
- [x] Password Min Age: 0 **(No minimum time between password changes)**
- [x] Password Min Length: 10 **(passwords are stored as a hash - therefore this is not relevant)**
- [x] Password Allow User Change: FALSE **(users have to use self-service portal to change their password)**

Admin password policy (**let's rename this to service account policy?**)
- [x] Non-expiry passwords for service accounts

Scenarios Tested:

- [x] Users/Admins/Service accounts can change passwords in self-service
- [x] Admin passwords expire (tested by setting expiry to 20 seconds)
- [x] User passwords expire (tested by setting expiry to 20 seconds)
- [x] Service accounts do not expire.
- [x] Admin/User locked out after 5 invalid password attempts within a 30 minutes period.
- [x] User/Admin can not set the same password through ldap-phpadmin.
- [x] Users cannot set password to previously used passwords
- [x] Standard user can not change LDAP policies

** The following tests are irrelevant as pwm binds using the user DN.**
- [ ] User's cannot change their own passwords.
- [ ] Admin cannot change their own password.
- [ ] Service accounts can not change their own passwords.

We also need to consider enforcing password complexity.

SachinKSingh28 committed:
Added password checker

Documentation updated to include password policy details and cleanup in dockerfile

SachinKSingh28 commented:
Thanks @RobertNorthard for testing this PR. I have tested some more scenarios -

Test cases :

- **Default password policy** :
	- [x] Password Attribute: userPassword
	- [x] Password failure count interval: 30 
	- [x] Password Grace Auth N Limit: 5 
	- [x] Password in history: 5 
	- [x] Password Lockout: TRUE
	- [x] Password Expiry Warning: 259200 (3 days) **Throws message in the system logs.**
	- [x] Password Lockout duration: 1800 (30 minutes)
	- [x] Password Max Age: 7776000 (90 days) 
	- [x] Password Max Failures: 5
	- [x] Password Min Age: 0
	- [x] Password Min Length: 10 **(passwords are stored as a hash - therefore this is not relevant)**
	- [x] Password Allow User Change: TRUE **(users have to use self-service portal to change their password)**

- **Service account password policy**
	- [x] Non-expiry passwords for service accounts

- **Password Complexity Policy** : 
    - [x] Minimum length of the password is 9.
    - [x] Atleast contains 1 uppercase letter.
    - [x] Atleast contains 6 lowercase letter.
    - [x] Atleast contains 1 digit letter.
    - [x] Atleast contains 1 punctation. 

- **Scenarios Tested** :
	- **New ldap installation** :
		- [x] Users/Service accounts can change passwords in self-service. (Expected : Yes, based on password complexity set by policy checker)
		- [x] Admin passwords expire (Expected: doesn't expire)
		- [x] User passwords expire (Expected: expires but it has to be explicitly changed after this image rolled out)
		- [x] Service accounts do not expire. (Expected: doesn't expire)
		- [x] User/Service accounts locked out after 5 invalid password attempts within a 30 minutes period. (Expected: expires after 5 attempts and unlocked after 30 minutes)
		- [x] User/Service accounts can not set the same password through ldap-phpadmin. (Expected : Not able to set atleast last five passwords.)
		- [x] Users cannot set password to previously used passwords. (Expected : Not able to set atleast last five passwords.)
		- [x] Standard user can not change LDAP policies. (Expected: No privileges to change policies.)

	- **Existing ldap installation** :
		- [x] Users/Service accounts can change passwords in self-service. (Expected : Yes, based on password complexity set by policy checker)
		- [x] Admin passwords expire (Expected: doesn't expire)
		- [x] User passwords expire (Expected: expires but it has to be explicitly changed after this image rolled out)
		- [x] Service accounts expire. (Expected: doesn't expire)
		- [x] User/Service accounts locked out after 5 invalid password attempts within a 30 minutes period. (Expected: expires after 5 attempts and unlocked after 30 minutes)
		- [x] User/Service accounts can not set the same password through ldap-phpadmin. (Expected : Not able to set atleast last five passwords.)
		- [x] Users cannot set password to previously used passwords. (Expected : Not able to set at least last five passwords.)
		- [x] Password not restored for existing users from platform secrets file. (Expected : Not restored)
		- [x] Password not expired for existing users. (Expected : doesn't expire so that existing users are not automatically locked out)
		- [x] Standard user can not change LDAP policies. (Expected: No privileges to change policies.)
		- [x] ~~Existing users have policy associated.~~ (Expected: should have a default policy) - **Doesn't work hence has to be associated manually**.
		- [x] ~~Users able to change the password after it is expired.~~ (Expected : they should be able to) - **Doesn't work hence only administrators can change the password after expiry**.

- **adop docker compose scenarios** : 
    - Password generator should comply with the policies.
    - People provide the password provisioning script validates it against the policy.
    - Jobs adding users with non complying passwords.
    - Root DN password must also comply.

**Need upgrade notes**

RobertNorthard commented:
LGTM. Thanks @SachinKSingh28. Let's raise the following issues for adop-docker-compose:

- Password generator should comply with the policies.
- People provide the password provisioning script validates it against the policy.
- Jobs adding users with non-complying passwords.
- Root DN password must also comply.

We also need to add password expiry notifications.
